File: src/commands/publish-to-curse-forge.ts
import {createReadStream, readFileSync} from 'fs'
import {basename} from 'path'
import {Command, InvalidArgumentError, Option} from 'commander'
import FormData from 'form-data'
import fetch from 'node-fetch'

export async function publishToCurseForge(args: string[]): Promise<void> {
  const opts = parseArgs(args)

  const data = new FormData()
  data.append('file', createReadStream(opts.file), basename(opts.file))
  data.append('metadata', JSON.stringify(opts.metadata))

  const response = await fetch(`https://minecraft.curseforge.com/api/projects/${opts.projectId}/upload-file`, {
    body: data,
    method: "POST",
    headers: {
      "X-Api-Token": opts.token,
    },
  })
  // eslint-disable-next-line no-console
  console.log(await response.text())
  process.exit(response.ok ? 0 : 1)
}

interface ParsedOptions {
  // request information
  file: string,
  token: string,
  projectId: number,

  metadata: {
    changelog: string,
    changelogType: 'html' | 'text' | 'markdown',
    displayName?: string,
    parentFileID?: number,
    gameVersions?: number[],
    releaseType: 'alpha' | 'beta' | 'release',
  }
}

function parseArgs(args: string[]): ParsedOptions {
  interface RealArgs {
    readonly file: string,
    readonly token: string,
    readonly projectId: number,

    readonly parentFile?: number,
    readonly name?: string,
    readonly changelog?: string,
    readonly changelogFile?: string,
    readonly changelogType: 'html' | 'text' | 'markdown',
    readonly releaseType: 'alpha' | 'beta' | 'release',
    readonly gameVersions: number[],
  }

  const opts = new Command()
    .requiredOption('-f, --file <resource file>', 'the path to jar file')
    .requiredOption('-t, --token <token>', 'the API token')
    .requiredOption('-i, --project-id <id>', 'the project id', numberParser('project id'))

    // metadata options
    .option('-p, --parent-file <id>', 'the parent file id', numberParser('parent file id'))
    .option('-n, --name <display name>', 'the display name of file')
    .option('--changelog <changelog text>', 'the changelog')
    .option('--changelog-file <changelog file>', 'the path to changelog')
    .addOption(new Option('--changelog-type <changelog file>', 'the document type of changelog')
      .default('html')
      .choices(['html', 'text', 'markdown']))
    .addOption(new Option('--release-type <release type>', 'the path to changelog file')
      .makeOptionMandatory()
      .choices(['alpha', 'beta', 'release']))
    .option('--game-versions <version id...>', 'the path to changelog file', parseVersionId, [])
    .parse(args, {from: 'user'})
    .opts<RealArgs>()

  if (opts.changelogFile && opts.changelog) {
    process.stderr.write("can't set both --changelog and --changelog-file")
    process.exit(1)
  }
  if (!(opts.changelogFile || opts.changelog)) {
    process.stderr.write("please set either --changelog or --changelog-file")
    process.exit(1)
  }

  const changelog: string = opts.changelogFile
    ? readFileSync(opts.changelogFile, {encoding: "utf8"})
    : opts.changelog as string
  

  return {
    file: opts.file,
    token: opts.token,
    projectId: opts.projectId,
    metadata: {
      changelog,
      changelogType: opts.changelogType,
      displayName: opts.name,
      parentFileID: opts.parentFile,
      gameVersions: opts.parentFile ? undefined : opts.gameVersions,
      releaseType: opts.releaseType,
    },
  }
}

function parseVersionId(value: string, previous: number[]): number[] {
  previous = previous || []
  const number = parseInt(value)
  if (Number.isNaN(number))
    throw new InvalidArgumentError(`invalid version id: ${value}`)
  return [...previous, number]
}

function numberParser(name: string): (value: string) => number {
  return (value) => {
    const number = parseInt(value)
    if (Number.isNaN(number))
      throw new InvalidArgumentError(`invalid ${name}: ${value}`)
    return number
  }
}

File: src/commands/publish_to_curse_forge.rs
use crate::CmdResult;
use clap::{Args, Parser, ValueEnum};
use reqwest::multipart::Part;
use serde::Serialize;
use std::num::NonZeroU64;
use std::path::PathBuf;
use tokio::fs;

#[derive(Debug, Parser)]
#[command(name = "publish-to-curse-forge")]
#[command(no_binary_name = true)]
pub(crate) struct PublishToCurseForge {
    // required options
    /// The path to the jar file to upload
    #[arg(short, long)]
    file: PathBuf,
    /// The API Token for curse forge
    #[arg(short, long)]
    token: String,
    /// The project id of the project to upload to
    #[arg(short = 'i', long)]
    project_id: u64,
    /// The release kind
    #[arg(long)]
    release_type: ReleaseType,

    // metadata options
    /// The parent file id
    #[arg(short = 'p', long)]
    parent_file_id: Option<NonZeroU64>,
    /// The display name of the file
    #[arg(short = 'n', long)]
    name: Option<String>,
    #[command(flatten)]
    changelog: ChangelogGroup,
    /// the document format of the changelog
    #[arg(long, default_value = "html")]
    changelog_type: ChangelogType,

    /// The game versions compatible with
    #[arg(long, value_delimiter = ',', num_args = 1..)]
    game_versions: Vec<u64>,
}

#[derive(Debug, Args)]
#[group(required = true, multiple = false)]
struct ChangelogGroup {
    /// The changelog text
    #[arg(long)]
    changelog: Option<String>,
    /// Path to the changelog file
    #[arg(long)]
    changelog_file: Option<PathBuf>,
}

#[derive(ValueEnum, Debug, Copy, Clone, Serialize)]
#[serde(rename_all = "lowercase")]
enum ReleaseType {
    Alpha,
    Beta,
    Release,
}

#[derive(ValueEnum, Debug, Copy, Clone, Default, Serialize)]
#[serde(rename_all = "lowercase")]
enum ChangelogType {
    #[default]
    Html,
    Text,
    Markdown,
}

#[derive(Serialize)]
struct CurseMetadata<'a> {
    changelog: String,
    changelog_type: ChangelogType,
    display_name: Option<&'a str>,
    parent_file_id: Option<NonZeroU64>,
    game_versions: &'a [u64],
    release_type: ReleaseType,
}

impl PublishToCurseForge {
    pub async fn run(&self) -> CmdResult {
        let in_file = fs::File::open(&self.file).await.expect("opening file");

        let client = reqwest::Client::new();
        let mut form = reqwest::multipart::Form::new();
        form = form.part("file", Part::stream(in_file));
        form = form.text(
            "metadata",
            serde_json::to_string(&self.metadata().await).unwrap(),
        );

        let response = client
            .post(format!(
                "https://minecraft.curseforge.com/api/projects/{}/upload-file",
                self.project_id
            ))
            .header("X-Api-Token", self.token.clone())
            .multipart(form)
            .send()
            .await
            .unwrap();
        let status = response.status();
        let text = response.text().await.unwrap();
        println!("{}", text);
        if status.is_success() {
            ok!()
        } else {
            err!()
        }
    }

    async fn metadata(&self) -> CurseMetadata<'_> {
        let changelog = match (&self.changelog.changelog, &self.changelog.changelog_file) {
            (Some(changelog), None) => changelog.clone(),
            (None, Some(changelog_file)) => fs::read_to_string(changelog_file)
                .await
                .expect("reading changelog"),
            (None, None) => panic!("no changelog"),
            (Some(_), Some(_)) => panic!("both changelog and changelog_file"),
        };
        CurseMetadata {
            changelog,
            changelog_type: self.changelog_type,
            display_name: self.name.as_deref(),
            parent_file_id: self.parent_file_id,
            game_versions: &self.game_versions,
            release_type: self.release_type,
        }
    }
}
